core/src/main/java/com/linecorp/armeria/client/ClientOptions.java

```diff
@@ -192,19 +177,70 @@ private static HttpHeaders filterHttpHeaders(HttpHeaders headers) {
     }
 
     private ClientOptions(ClientOptionValue<?>... options) {
-        super(ClientOptions::filterValue, options);
+        super(options);
+    }
+
+    private ClientOptions(ClientOptions baseOptions, ClientOptionValue<?>... additionalOptions) {
+        super(baseOptions, additionalOptions);
     }
 
-    private ClientOptions(ClientOptions clientOptions, ClientOptionValue<?>... options) {
-        super(ClientOptions::filterValue, clientOptions, options);
+    private ClientOptions(ClientOptions baseOptions, Iterable<ClientOptionValue<?>> additionalOptions) {
+        super(baseOptions, additionalOptions);
     }
 
-    private ClientOptions(ClientOptions clientOptions, Iterable<ClientOptionValue<?>> options) {
-        super(ClientOptions::filterValue, clientOptions, options);
+    private ClientOptions(ClientOptions baseOptions, ClientOptions additionalOptions) {
+        super(baseOptions, additionalOptions);
     }
 
-    private ClientOptions(ClientOptions clientOptions, ClientOptions options) {
-        super(clientOptions, options);
+    @Override
+    protected <T extends AbstractOptionValue<?, ?>> T filterValue(T optionValue) {
+        if (optionValue.option() == HTTP_HEADERS) {
+            @SuppressWarnings("unchecked")
+            final ClientOption<HttpHeaders> castOption = (ClientOption<HttpHeaders>) optionValue.option();
+            @SuppressWarnings("unchecked")
+            final T castOptionValue = (T) castOption.newValue(
+                    filterHttpHeaders((HttpHeaders) optionValue.value()));
+            return castOptionValue;
+        }
+        return optionValue;
+    }
+
+    @Override
+    protected <T extends AbstractOptionValue<?, ?>> T mergeValue(T oldValue, T newValue) {
+        if (oldValue.option() == DECORATION) {
+            final ClientDecoration oldDecoration = (ClientDecoration) oldValue.value();
+            final ClientDecoration newDecoration = (ClientDecoration) newValue.value();
+            if (oldDecoration.decorators().isEmpty() && oldDecoration.rpcDecorators().isEmpty()) {
```

Comment:
How about adding `ClientDecoration.isEmpty()` and simplifying here?

Reply:
Oh, that sounds nice.